service_container/injection_types.rst

```diff
@@ -100,6 +100,112 @@ working with optional dependencies. It is also more difficult to use in
+        }
+
+        // ...
+    }
+
+In order to use the full potential of this approach, you can define a third argument
+which allows the container to return the newly created service:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/services.yaml
+       services:
+            # ...
+
+            app.newsletter_manager:
+                class: App\Mail\NewsletterManager
+                calls:
+                    - [withMailer, ['@mailer'], true]
+
+    .. code-block:: xml
+
+        <!-- config/services.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+                <!-- ... -->
+
+                <service id="app.newsletter_manager" class="App\Mail\NewsletterManager">
+                    <call method="withMailer" use-result="true">
+                        <argument type="service" id="mailer" />
+                    </call>
+                </service>
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        // config/services.php
+        use App\Mail\NewsletterManager;
+        use Symfony\Component\DependencyInjection\Reference;
+
+        // ...
+        $container->register('app.newsletter_manager', NewsletterManager::class)
+            ->addMethodCall('withMailer', [new Reference('mailer')], true);
+
+.. note::
+
+    This type of injection requires that you add a ``@return static`` docblock in order
+    for the container to be capable of registering the method.
+
+This approach is useful if you need to configure your service according to your needs,
+so, here's the advantages of immutable-setters:
+
+* Immutable setters works with optional dependencies, this way, if you don't need
+  a dependency, the setter don't need to be called.
+
+* Like the constructor injection, using immutable setters force the dependency to stay
+  the same during the lifetime of a service.
+
+* Your service becomes immutable, as the container will return a new object,
+  the initial service stays clean and unchanged.
+
+* You can easily change the injected service as long as it respect the interface/type
+  asked by the initial service.
+
+The disadvantages are:
+
+* As the setter call is optional, a dependency can be null during execution,
+  you must check that the dependency is available before calling it.
+
+* As the ``@return static`` docblock is required by the container to
+  understand that the method return a new object,
+  you can found that adding docblock for a single method isn't adapted or
+  link your code to the container.
+
+* As this approach force the container to create a new object
+  once the method is called, you can found hard to debug and test your code.
+
```

Comment:
The whole new docs part should be below the setter injection part, what do you think?

Reply:
It depends, @nicolas-grekas can probably answer this question better than me, does the "immutable-setter" injection is recommended over the setter one?